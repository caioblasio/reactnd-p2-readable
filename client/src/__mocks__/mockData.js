const testPosts = {

  "8xf0y6ziyjabvozdd253nd": {
    id: '8xf0y6ziyjabvozdd253nd',
    timestamp: 1467166872634,
    title: 'Udacity is the best place to learn React',
    body: 'Everyone says so after all.',
    author: 'thingtwo',
    category: 'categoryone',
    voteScore: 6,
    deleted: false,
    commentCount: 2
  },
  "6ni6ok3ym7mf1p33lnez": {
    id: '6ni6ok3ym7mf1p33lnez',
    timestamp: 1468479767190,
    title: 'Learn Redux in 10 minutes!',
    body: 'Just kidding. It takes more than 10 minutes to learn technology.',
    author: 'thingone',
    category: 'categorytwo',
    voteScore: -5,
    deleted: false,
    commentCount: 0
  }
}

const testCategories = [
  {
    name: 'Category One',
    path: 'categoryone'
  },
  {
    name: 'Category Two',
    path: 'categorytwo'
  }
]

const testComments = [
  {
    id: '894tuq4ut84ut8v4t8wun89g',
    parentId: "8xf0y6ziyjabvozdd253nd",
    timestamp: 1468166872634,
    body: 'Hi there! I am a COMMENT.',
    author: 'thingtwo',
    voteScore: 6,
    deleted: false,
    parentDeleted: false
  },
  {
    id: '8tu4bsun805n8un48ve89',
    parentId: "8xf0y6ziyjabvozdd253nd",
    timestamp: 1469479767190,
    body: 'Comments. Are. Cool.',
    author: 'thingone',
    voteScore: -5,
    deleted: false,
    parentDeleted: false
  }
  
]

/* JSON headers for API requests */
const jsonHeaders = new Headers({
	Accept: 'application/json',
	'Content-Type': 'application/json'
});

export { testPosts, testCategories, testComments, jsonHeaders };