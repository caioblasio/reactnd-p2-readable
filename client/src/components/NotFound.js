import React, { Component } from 'react';

import { withRouter } from 'react-router-dom';

import { withStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';


import Typography from '@material-ui/core/Typography';

import Divider from '@material-ui/core/Divider';
import CardActions from '@material-ui/core/CardActions';
import Button from '@material-ui/core/Button';

const styles = theme => ({
  card: {
    width: '70%',
    margin: '0 auto',
  },
  link: {
    ...theme.link,
    outline: 0,
  },
  actions: {
    display: 'flex',
    alignItems: 'center',
  },
});

class NotFound extends Component {

  render() {

    const { classes, history } = this.props;
    return (
      <Card className={classes.card}>
          <CardHeader
            title="Page Not Found :("
          />
          <CardContent>
            <Typography component="p">
                Maybe the page you are looking for has been removed, or you typed in the wrong URL
            </Typography>
          </CardContent>
          <CardActions className={classes.actions} disableActionSpacing>
            <Button size="small" color="primary" onClick={() => {history.goBack()}}>
              Go Back
            </Button>
            <Button size="small" color="primary" onClick={() => {history.push('/')}}>
              Go To Homepage
            </Button>
          </CardActions>
      </Card>
    )
  }
}

export default withStyles(styles)(withRouter(NotFound))