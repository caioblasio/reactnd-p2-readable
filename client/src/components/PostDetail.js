import React, { Component } from 'react';
import { connect } from 'react-redux';

import { fetchPostById } from '../actions/posts';

class PostDetail extends Component {

  componentDidMount(){
    const { dispatch, id } = this.props;
    dispatch(fetchPostById(id))
  }

  render() {
    const { post } = this.props;
    return (
      <div>
        {post && 
          <div>{post.title}</div>
        }
      </div>
    )
  }
}

function mapStateToProps({ posts }, {id}) {
  return {
    post: posts[id]
  }
}

export default connect(mapStateToProps)(PostDetail)