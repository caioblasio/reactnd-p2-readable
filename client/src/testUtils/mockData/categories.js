export const testCategories = [
  {
    name: 'Category One',
    path: 'categoryone'
  },
  {
    name: 'Category Two',
    path: 'categorytwo'
  }
]