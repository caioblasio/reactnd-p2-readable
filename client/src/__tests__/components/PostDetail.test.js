import React from 'react';
import PostDetail from '../../components/PostDetail';
import configureStore from 'redux-mock-store';
import toJson from 'enzyme-to-json';
import { testPosts, testComments, jsonHeaders } from '../../__mocks__/mockData';
import { MemoryRouter, Router } from 'react-router-dom';
import thunk from 'redux-thunk'
import NotFound from '../../components/NotFound';
import { Provider } from 'react-redux';
import IconButton from '@material-ui/core/IconButton';

import { createShallow } from '@material-ui/core/test-utils';

const middlewares = [ thunk ];
const mockStore = configureStore(middlewares);

describe('<PostDetail />', () => {

  let wrapper, store;


  beforeAll(() => {

    const initialState = {
      posts: testPosts,
      comments: testComments,
      loading: false
    }

    store = mockStore(initialState);

  });

  beforeEach(() => {
    fetch.resetMocks();
  });

  it('renders the component', () => {

    const testPost = testPosts[Object.keys(testPosts)[0]];

    fetch
      .mockResponse(JSON.stringify(testPost), { jsonHeaders })

    wrapper = shallow(
      <PostDetail
        id={testPost.id}
        store={store}
      />
    );
    const component = wrapper.dive();
    expect(toJson(component.dive())).toMatchSnapshot();
  });

  it('should render a Not Found', () => {

    const error = { error: 'There was an error' }

    fetch
      .mockResponse(JSON.stringify(error), { jsonHeaders })

    wrapper = shallow(
      <PostDetail
        id="0123456789" //id that not exist
        store={store}
      />
    );
    const component = wrapper.dive();
    expect(component.dive().find(NotFound)).toHaveLength(1);
  });

  it('should render a Correct Post', () => {

    const testPost = testPosts[Object.keys(testPosts)[0]];

    fetch
      .mockResponse(JSON.stringify(testPost), { jsonHeaders })

    wrapper = shallow(
      <PostDetail
        id={testPost.id}
        store={store}
      />
    );
    const component = wrapper.dive();
    expect(component.props().post.id).toBe(testPost.id);
    
  });

  test('should delete a post', (done) => {

    let shallow = createShallow({untilSelector: PostDetail})

    const testPost = testPosts[Object.keys(testPosts)[0]];
    const id = testPost.id
    window.confirm = jest.fn(() => true);

    fetch
      .once(JSON.stringify(testPost))
      .once(JSON.stringify({ id }))

   

    wrapper = shallow(
      <MemoryRouter initialEntries={[`/post/${id}`]}>
        <PostDetail
          id={id}
          store={store}
          history={[]}
        />
      </MemoryRouter>
    );

    const component = wrapper.dive();
    component.dive().find(IconButton).last().simulate('click');

    const expectedAction = { type: 'REMOVE_POST', id: '8xf0y6ziyjabvozdd253nd' }

    setImmediate(() => {
      expect(store.getActions()).toEqual(
        expect.arrayContaining([expectedAction])
      );
      done();
    })
  });



});