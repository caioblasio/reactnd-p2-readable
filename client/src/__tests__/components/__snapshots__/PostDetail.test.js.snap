// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostDetail /> renders the component 1`] = `
<Fragment>
  <div
    className="Connect-PostDetail--root-1"
  >
    <div
      className="Connect-PostDetail--post-2"
    >
      <WithStyles(Paper)
        className="Connect-PostDetail--detail-3"
        elevation={1}
      >
        <WithStyles(Typography)
          color="default"
          variant="h5"
        >
          Udacity is the best place to learn React
        </WithStyles(Typography)>
        <WithStyles(Typography)
          className="Connect-PostDetail--headline-5"
          color="default"
          variant="body1"
        >
          by 
          <strong>
            thingtwo
          </strong>
          <span
            className="Connect-PostDetail--headlineItems-6"
          >
            Jun 28, 2016
          </span>
          <strong>
            categoryone
          </strong>
        </WithStyles(Typography)>
        <WithStyles(Typography)
          className="Connect-PostDetail--body-7"
          color="default"
          variant="body1"
        >
          Everyone says so after all.
        </WithStyles(Typography)>
        <WithStyles(Divider)
          className="Connect-PostDetail--divider-8"
        />
        <div
          className="Connect-PostDetail--action-9"
        >
          <div
            className="Connect-PostDetail--vote-10"
          >
            <Connect(VoteControl)
              id="8xf0y6ziyjabvozdd253nd"
              type="post"
              voteScore={6}
            />
          </div>
          <div>
            <WithStyles(IconButton)
              aria-label="Like"
              onClick={[Function]}
            >
              <Link
                className="Connect-PostDetail--link-11"
                replace={false}
                to="/edit/8xf0y6ziyjabvozdd253nd"
              >
                <pure(EditIcon) />
              </Link>
            </WithStyles(IconButton)>
            <WithStyles(IconButton)
              aria-label="Like"
              onClick={[Function]}
            >
              <pure(DeleteIcon) />
            </WithStyles(IconButton)>
          </div>
        </div>
      </WithStyles(Paper)>
      <WithStyles(Paper)
        className="Connect-PostDetail--detail-3"
        elevation={1}
      >
        <WithStyles(Connect(CommentList))
          postId="8xf0y6ziyjabvozdd253nd"
        />
      </WithStyles(Paper)>
      <WithStyles(Paper)
        className="Connect-PostDetail--detail-3"
        elevation={1}
      >
        <WithStyles(Connect(NewComment))
          parentId="8xf0y6ziyjabvozdd253nd"
        />
      </WithStyles(Paper)>
    </div>
    <div
      className="Connect-PostDetail--side-4"
    >
      <WithStyles(Connect(PostList))
        category="categoryone"
        excludeId="8xf0y6ziyjabvozdd253nd"
        headline="Related Posts"
        maxQty={4}
        side={true}
      />
    </div>
  </div>
</Fragment>
`;
